ios/objc/src/TorchTensor.mm

```diff
@@ -0,0 +1,196 @@
+#import "TorchTensor.h"
+#import "TorchTensor+Internal.h"
+#import <Pytorch/Pytorch.h>
+
+#define CHECK_IMPL(x) NSCAssert(x!=nil,@"impl is nil!");
+#define CHECK_IMPL_(x) \
+    CHECK_IMPL(x) \
+    if (!x) { return nil; }
+
+#define DEFINE_TENSOR_TYPES(_) \
+    _(Byte) \
+    _(Int) \
+    _(Float) \
+    _(Long) \
+    _(Undefined)
+
+#define DEFINE_TENSOR_SCALAR_TYPES(_) \
+    _(Int) \
+    _(Float) \
+    _(Long) \
+
+static inline c10::ScalarType scalarTypeFromTensorType(TorchTensorType type) {
+    switch(type){
+#define DEFINE_CASE(x) case TorchTensorType##x: return c10::ScalarType::x;
+        DEFINE_TENSOR_TYPES(DEFINE_CASE)
+#undef DEFINE_CASE
+    }
+    return c10::ScalarType::Undefined;
+}
+
+static inline TorchTensorType tensorTypeFromScalarType(c10::ScalarType type) {
+    switch(type){
+#define DEFINE_CASE(x) case c10::ScalarType::x: return TorchTensorType##x;
+        DEFINE_TENSOR_TYPES(DEFINE_CASE)
+#undef DEFINE_CASE
+        default: return TorchTensorTypeUndefined;
+    }
+}
+
+@implementation TorchTensor {
+    std::shared_ptr<at::Tensor> _impl;
+}
+
+- (TorchTensorType)type {
+    CHECK_IMPL(_impl)
+    if(!_impl){
+        return TorchTensorTypeUndefined;
+    }
+    return tensorTypeFromScalarType(_impl->scalar_type());
+}
+
+- (BOOL)quantized{
+    return _impl ? _impl->is_quantized() : NO;
+}
+
+- (int64_t)numel{
+    CHECK_IMPL(_impl);
+    if(!_impl) {
+        return NSNotFound;
+    }
+    return _impl->numel();
+}
+
+- (void* )data {
+    CHECK_IMPL_(_impl);
+    return _impl->unsafeGetTensorImpl()->storage().data();
+}
+
+- (int64_t)dim {
+    CHECK_IMPL(_impl);
+    if(!_impl) {
+        return NSNotFound;
+    }
+    return _impl->dim();
+}
+
++ (TorchTensor* )newWithType:(TorchTensorType)type Size:(NSArray<NSNumber* >*)size Data:(void* _Nullable)data {
+    return [self newWithType:type Size:size Data:data Quantized:NO];
+}
+
+
++ (TorchTensor* )newWithType:(TorchTensorType)type Size:(NSArray<NSNumber* >*)size Data:(void* _Nullable)data Quantized:(BOOL) quantized {
+    if (!data || size.count == 0){
+        return nil;
+    }
+    std::vector<int64_t> dimsVec;
+    for (auto i = 0; i < size.count; ++i) {
+        int64_t dim = size[i].integerValue;
+        dimsVec.push_back(dim);
+    }
+    at::Tensor tensor = torch::from_blob( (void* )data, dimsVec,scalarTypeFromTensorType(type));
```

Comment:
here and everywhere: exception safety. We probably should catch exception (all of Torch throws c10::Exception) and wrap it into what ObjC exception mechanism is

Reply:
May simply log the error and return `nil`, since that's a constructor method, and returning `nil` is a typical wait to convey failed construction.
Exceptions are sorta deprecated in ObjC for 3rd party devs. Usually there's an out arg of type `NSError` to indicate the method can fail, and sometimes a `BOOL` return value in addition, to indicate success or failure.